File: bench/cs/Ssh.Benchmark/PortForwardBenchmark.cs
using System;
using System.Diagnostics;
using System.Net;
using System.Net.Sockets;
using System.Security.Claims;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Events;
using Microsoft.DevTunnels.Ssh.Messages;
using Microsoft.DevTunnels.Ssh.Tcp;
using Microsoft.DevTunnels.Ssh.Tcp.Events;

namespace Microsoft.DevTunnels.Ssh.Benchmark;

#if NETSTANDARD2_0 || NET4
using ValueTask = System.Threading.Tasks.Task;
#endif

class PortForwardBenchmark : Benchmark
{
	private const string ConnectTimeMeasurement = "Connect time (ms)";

	private readonly IPAddress listenAddress;
	private readonly string hostAddress;

	private readonly SshServer server;
	private readonly Task serverTask;
	private readonly SshClient client;

	public PortForwardBenchmark(IPAddress listenAddress, string hostAddress)
		: base($"Port forward to {hostAddress} ({listenAddress})")
	{
		HigherIsBetter[ConnectTimeMeasurement] = false;

		this.listenAddress = listenAddress;
		this.hostAddress = hostAddress;

		var config = new SshSessionConfiguration();
		config.AddService(typeof(PortForwardingService));

		var trace = new TraceSource(nameof(PortForwardBenchmark));
		this.server = new SshServer(config, trace);
		this.client = new SshClient(config, trace);

		var serverKey = SshAlgorithms.PublicKey.RsaWithSha512.GenerateKeyPair();
		this.server.Credentials = new[] { serverKey };

		this.server.SessionAuthenticating += OnServerSessionAuthenticating;

		this.serverTask = this.server.AcceptSessionsAsync(Benchmark.ServerPort, IPAddress.Loopback);
	}

	protected override async Task RunAsync(Stopwatch stopwatch)
	{
		var serverSessionCompletion = new TaskCompletionSource<SshServerSession>();
		EventHandler<SshServerSession> sessionOpenedHandler = null;
		sessionOpenedHandler = (sender, session) =>
		{
			serverSessionCompletion.SetResult(session);
			this.server.SessionOpened -= sessionOpenedHandler;
		};
		this.server.SessionOpened += sessionOpenedHandler;

		var clientSession = await this.client.OpenSessionAsync(
			IPAddress.Loopback.ToString(), ServerPort);
		clientSession.Authenticating += OnClientSessionAuthenticating;
		clientSession.Request += OnClientSessionRequest;
		await clientSession.AuthenticateAsync(("benchmark", "benchmark"));
		var serverSession = await serverSessionCompletion.Task;

		var connectServer = new TcpListener(this.listenAddress, 0);
		connectServer.Start();
		var serverPort = ((IPEndPoint)connectServer.LocalEndpoint).Port;

		var availablePortListener = new TcpListener(IPAddress.Loopback, 0);
		availablePortListener.Start();
		var clientPort = ((IPEndPoint)availablePortListener.LocalEndpoint).Port;
		availablePortListener.Stop();

		var forwarder = await serverSession.ForwardFromRemotePortAsync(
			IPAddress.Loopback, clientPort, this.hostAddress, serverPort);
		if (forwarder == null)
		{
			throw new InvalidOperationException("Failed to forward port");
		}

		var channelOpenedCompletion = new TaskCompletionSource<ForwardedPortChannelEventArgs>();
		clientSession.GetService<PortForwardingService>().RemoteForwardedPorts.PortChannelAdded +=
			(sender, e) => channelOpenedCompletion.SetResult(e);

		var connectClient = new TcpClient();
		double connectStartMark = stopwatch.Elapsed.TotalMilliseconds;

		var clientConnectTask = connectClient.ConnectAsync(IPAddress.Loopback, clientPort);
		var serverAcceptTask = connectServer.AcceptTcpClientAsync();
		await Task.WhenAll(clientConnectTask, serverAcceptTask, channelOpenedCompletion.Task);
		double connectEndMark = stopwatch.Elapsed.TotalMilliseconds;

		AddMeasurement(ConnectTimeMeasurement, connectEndMark - connectStartMark);
	}

	private void OnClientSessionAuthenticating(object sender, SshAuthenticatingEventArgs e)
	{
		e.AuthenticationTask = Task.FromResult(new ClaimsPrincipal());
	}

	private void OnServerSessionAuthenticating(object sender, SshAuthenticatingEventArgs e)
	{
		e.AuthenticationTask = Task.FromResult(new ClaimsPrincipal());
	}

	private void OnClientSessionRequest(object sender, SshRequestEventArgs<SessionRequestMessage> e)
	{
		e.IsAuthorized = true;
	}

	public override async ValueTask DisposeAsync()
	{
		this.server.Dispose();
		this.client.Dispose();
		await this.serverTask;
	}
}

File: bench/ts/ssh-bench/portForwardBenchmark.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as net from 'net';
import { Benchmark } from './benchmark';
import {
	SshClientCredentials,
	SshSessionConfiguration,
	SshAlgorithms,
	PromiseCompletionSource,
	SshServerSession,
} from '@microsoft/dev-tunnels-ssh';
import {
	ForwardedPortChannelEventArgs,
	PortForwardingService,
	SshClient,
	SshServer,
} from '@microsoft/dev-tunnels-ssh-tcp';

const ConnectTimeMeasurement = 'Connect time (ms)';

declare type hrtime = [number, number];
const millis = ([s, ns]: hrtime) => s * 1000 + ns / 1000000;

export class PortForwardBenchmark extends Benchmark {
	private readonly hostAddress: string;
	private readonly listenAddress: string;
	private readonly port: number;
	private readonly initPromise: Promise<void>;
	private readonly server: SshServer;
	private readonly serverPromise: Promise<void>;
	private readonly client: SshClient;

	public constructor(listenAddress: string, hostAddress: string) {
		super(`Port forward to ${hostAddress} (${listenAddress})`);

		this.higherIsBetter.set(ConnectTimeMeasurement, false);

		this.hostAddress = hostAddress;
		this.listenAddress = listenAddress;

		const config = new SshSessionConfiguration();
		config.addService(PortForwardingService);

		this.server = new SshServer(config);
		this.client = new SshClient(config);

		////this.server.trace = (_, __, msg) => console.log('SERVER: ' + msg);
		////this.client.trace = (_, __, msg) => console.log('CLIENT: ' + msg);

		this.initPromise = SshAlgorithms.publicKey
			.rsaWithSha512!.generateKeyPair()
			.then((serverKey) => {
				this.server.credentials.publicKeys.push(serverKey);
			});

		this.server.onSessionOpened((session) => {
			session.onAuthenticating((e) => {
				e.authenticationPromise = Promise.resolve({});
			});
		});

		this.port = Benchmark.findAvailablePort();
		this.serverPromise = this.server.acceptSessions(this.port, '127.0.0.1');
	}

	public async run(): Promise<void> {
		await this.initPromise;

		const startTime: hrtime = process.hrtime();

		const serverSessionCompletion = new PromiseCompletionSource<SshServerSession>();
		const sessionOpenedRegistration = this.server.onSessionOpened((session) => {
			serverSessionCompletion.resolve(session);
			sessionOpenedRegistration.dispose();
		});

		var clientSession = await this.client.openSession('127.0.0.1', this.port);
		clientSession.onAuthenticating((e) => {
			e.authenticationPromise = Promise.resolve({});
		});
		await clientSession.authenticate({ username: 'benchmark' });
		clientSession.onRequest((e) => {
			e.isAuthorized = true;
		});

		const serverSession = await serverSessionCompletion.promise;

		const connectServer = new net.Server();
		const listeningCompletion = new PromiseCompletionSource<void>();
		connectServer.listen(0, this.listenAddress, undefined, () => listeningCompletion.resolve());
		await listeningCompletion.promise;
		const serverPort = (<net.AddressInfo>connectServer.address()).port;

		const availableServer = new net.Server();
		const availableCompletion = new PromiseCompletionSource<void>();
		availableServer.listen(0, this.listenAddress, undefined, () => availableCompletion.resolve());
		await availableCompletion.promise;
		const clientPort = (<net.AddressInfo>availableServer.address()).port;
		await new Promise<void>((resolve, reject) => {
			availableServer.close((err) => (err ? reject(err) : resolve()));
		});

		const serverPfs = serverSession.activateService(PortForwardingService);
		const forwarder = await serverPfs.forwardFromRemotePort(
			'127.0.0.1',
			clientPort,
			this.hostAddress,
			serverPort,
		);
		if (!forwarder) throw new Error('Failed to forward port.');

		const channelOpenedCompletion = new PromiseCompletionSource<ForwardedPortChannelEventArgs>();
		const clientPfs = clientSession.activateService(PortForwardingService);
		clientPfs.remoteForwardedPorts.onPortChannelAdded((e) => channelOpenedCompletion.resolve(e));

		const connectClient = new net.Socket();
		const connectStartMark: hrtime = process.hrtime(startTime);

		const serverAcceptPromise = new Promise<void>((resolve, reject) => {
			connectServer.on('error', reject).once('connection', resolve);
		});
		const connectClientPromise = new Promise<void>((resolve, reject) => {
			connectClient.on('error', reject).connect(clientPort, '127.0.0.1', resolve);
		});
		await Promise.all([
			connectClientPromise,
			serverAcceptPromise,
			channelOpenedCompletion.promise,
		]);

		const connectEndMark: hrtime = process.hrtime(startTime);

		connectClient.end();
		await new Promise<void>((resolve, reject) => {
			connectServer.close((err) => (err ? reject(err) : resolve()));
		});

		this.addMeasurement(
			ConnectTimeMeasurement,
			millis(connectEndMark) - millis(connectStartMark),
		);
	}

	public async dispose(): Promise<void> {
		this.server.dispose();
		this.client.dispose();
		await this.serverPromise;
	}
}
